### PR TITLE
fix(cli): tighten scoreTypeFidelity flag-decl regex and drop MarkFlagRequired reward

### LINKED ISSUE #1559: scorer: scoreTypeFidelity rewards MarkFlagRequired which the SKILL explicitly forbids (verify-breaking)
> Filed by `/printing-press-retro` from a scorecard-walk follow-up on
> **Blu-ray.com** (final scorecard 81/100 Grade A; this finding accounts
> for -1 fixed point on a 5-point dim no agent following the SKILL can
> avoid).

## Summary

`scoreTypeFidelity` awards `+1` point when a CLI declares `MarkFlagRequired(...)`
at least three times across the 10 sampled command files. The SKILL's Phase 3
build checklist (item 8: "Verify-friendly RunE") explicitly **FORBIDS**
`MarkFlagRequired` on hand-written commands because it gates `--dry-run`
probes: Cobra evaluates `MarkFlagRequired` before `RunE` runs, so a
`dryRunOK(flags)` short-circuit inside `RunE` cannot reach if a required
flag is missing — verify's `--dry-run` probe then fails with `required flag
"X" not set`.

An agent that correctly follows the SKILL ships zero `MarkFlagRequired`
calls and loses the point. An agent that ignores the SKILL gains the
point and breaks verify. This is a direct scorer-versus-SKILL conflict.

## Where to look

- **Component:** scorer
- **Likely area:** `internal/pipeline/scorecard.go` `scoreTypeFidelity`
  (around line 2266, the `requiredCount >= 3` block).
- **Triggered when:** any CLI that correctly follows the SKILL's
  verify-friendly RunE pattern. Spec-generated commands typically don't
  use `MarkFlagRequired` either (the generator validates required flags
  in RunE so dry-run probes pass); hand-written commands explicitly must
  not. So this is a near-universal -1 against the dim cap.

## What we observed

`scorecard.go:2229` `scoreTypeFidelity`:

```go
flagDeclRe := regexp.MustCompile(`Flags\(\)\.(StringVar|...)\(...`)
requiredRe := regexp.MustCompile(`MarkFlagRequired\("([^"]+)"\)`)
...
requiredCount += len(requiredRe.FindAllStringSubmatch(content, -1))
...
if requiredCount >= 3 {
    score++
}
```

`skills/printing-press/SKILL.md` Phase 3 build checklist item 8:

> **Verify-friendly RunE**: Hand-written commands MUST NOT use `Args: cobra.MinimumNArgs(N)` or `MarkFlagRequired(...)`. Cobra evaluates both before RunE runs, so a `--dry-run` guard inside RunE cannot reach if those gates fail.

Blu-ray-pp-cli total `MarkFlagRequired` count: 1 (in framework files, not in
hand-written novel commands). `scoreTypeFidelity` reports 3/5 — would be
4/5 if the SKILL forbade what the scorer rewards (and 5/5 with one other
dim, but that's a separate path).

## Suspected root cause

The scorer was written before the SKILL's verify-friendly RunE rule
existed. The `MarkFlagRequired` signal was added as a proxy for "this CLI
exposes required flags cleanly" — but the SKILL has since established that
the correct way to handle required inputs is in-RunE validation, not
`MarkFlagRequired`. The scorer signal didn't migrate.

## Suggested direction

Two equivalent paths:

(a) **Drop the `MarkFlagRequired` signal entirely.** Replace with "hand-written
commands validate required inputs in RunE" — detectable as files where the
first non-comment statement in `RunE` is a guard returning `fmt.Errorf(...)`
on missing/empty input. This honors the SKILL's pattern and doesn't reward
the dry-run-breaking pattern.

(b) **Move the +1 to a different in-RunE validation signal.** Look for the
`if len(args) == 0 { return cmd.Help() }` pattern OR for explicit value
validation (`strconv.Atoi(args[0])` etc.) before any HTTP/IO calls. Any
positive-signal that aligns with the SKILL works.

Either path also removes the scorer's incentive for agents to ignore the
SKILL to bump their score.

## Acceptance criteria

- positive: regenerate a CLI following the SKILL's verify-friendly RunE
  pattern (zero `MarkFlagRequired`, validation in RunE) — assert the dim
  reaches whatever cap the new signal supports (e.g., +1 if the in-RunE
  guard is detected).
- negative: a CLI that DOES use `MarkFlagRequired` (and consequently fails
  verify's `--dry-run` probes) does NOT score higher on type-fidelity than
  one that follows the SKILL.

## Frequency

every API. Both spec-generated commands and hand-written commands following
the SKILL avoid `MarkFlagRequired`. The scorer-vs-SKILL conflict hits every
printed CLI uniformly.

## Complexity

small. One regex swap + a couple of tests in `scorecard_test.go`.

## Scope boundary

Does NOT touch the rest of `scoreTypeFidelity` (the ID-flag-as-string
check, the desc-word-count check, the blank-assign check are all
SKILL-aligned). Does NOT add new dim weight.

## Dependencies

None.

## Related issues

- #1457 — `Scorers under-score non-REST CLIs: hardcoded helpers.go/client.go greps miss the SKILL's sanctioned scaffolding-only path` — `related-area`. Same class of bug (scorer behavior contradicts the SKILL). Different specific signal; the structural fix here is independent.

## Artifacts

| Artifact | Link |
|----------|------|
| Retro document | https://files.catbox.moe/5updn7.md |
| Manuscripts (research + proofs) | https://files.catbox.moe/i7psqi.zip |
| Generated CLI source code | https://files.catbox.moe/rffoxy.zip |

---

*Generated by `/printing-press-retro` · [CLI Printing Press](https://github.com/mvanhorn/cli-printing-press)*

### LINKED ISSUE #1720: scoreTypeFidelity regex over-matches across newlines + 'id' substring false positive
> Filed by `/printing-press-retro` from a coffee-goat session (post-polish score-climb). No artifact uploads — findings are reproducible by running `printing-press scorecard` against any hand-authored CLI in the local library.

## Summary

`scoreTypeFidelity` in `internal/pipeline/scorecard.go` has two distinct defects in its flag-declaration detection. Both systematically penalize hand-authored CLIs even when their flags are well-shaped.

### Bug 1 — Regex over-matches across newlines

The package-level `flagDeclRe` regex:

```
Flags\(\)\.(StringVar|IntVar|StringVarP|IntVarP)\(&[^,]+,\s*"([^"]+)"(?:,\s*[^,]+){1,2},\s*"([^"]*)"
```

`[^,]+` matches across newlines and `{1,2}` is greedy. For consecutive `Flags()...` calls, the regex consumes "everything up to the next comma" — which spans into the *next* line's `Flags()...` invocation. The result: it captures the next flag's *name* (a kebab-case 1-3 word string) as the current flag's *description*, dragging the description-word-count average down even when every real description is ≥7 words.

### Bug 2 — "id" substring false positive

```go
if strings.Contains(name, "id") {
    totalIDFlags++
    if strings.HasPrefix(match[1], "StringVar") {
        stringIDFlags++
    }
}
```

`strings.Contains(name, "id")` has no word boundary. A flag named `price-paid-cents` matches because "paid" contains "id". An `IntVar` for cents is then classified as a non-string "ID flag," failing the "all ID flags must be StringVar" check and costing 2 points on the dimension.

## Where to look

- **Component:** scorer
- **File:** `internal/pipeline/scorecard.go`
- **Function:** `scoreTypeFidelity` (and the package-level `flagDeclRe`)

## Cross-CLI evidence

Reproduced offline by running the exact scorer regex against three hand-authored CLIs:

| CLI | Captured matches | Artifact "1-3 word kebab-case" descriptions | ID-flag false positives observed |
|-----|------------------|---------------------------------------------|----------------------------------|
| coffee-goat | 22 | ~8 (descriptions like `roast-date`, `purchase-date`, `mass-g`, `notes`, `time-s`, `temperature-c`, `water-tds-ppm`, `descriptors`) | 1 (`price-paid-cents` → matched "id" substring; required renaming the flag to `price-cents` to clear) |
| recipe-goat | 15 | 2 | 0 (no flag names with "id" substring in first-10 sample) |
| yahoo-fantasy | 8 | 3 | 0 |

Generator-emitted CLIs partially dodge bug 1 because flags are spread across separate functions (one per subcommand), so consecutive `Flags()` lines are rarer. Hand-authored CLIs with dense flag blocks (e.g., `beans.go` add-command with 7 sequential `Flags()` calls) hit it hardest.

## How an agent verified the bug

```python
import re
flag_re = re.compile(r'Flags\(\)\.(StringVar|IntVar|StringVarP|IntVarP)\(&[^,]+,\s*"([^"]+)"(?:,\s*[^,]+){1,2},\s*"([^"]*)"')
for f in first_10_non_infra_files:
    for m in flag_re.finditer(open(f).read()):
        kind, name, desc = m.group(1), m.group(2), m.group(3)
        # desc often looks like 'roast-date' — that's the NEXT flag's name, not a description
```

For coffee-goat's `beans.go` add-command block, this returned descriptions like `'product': 'roast-date'` (1 word) and `'mass-g': 'notes'` (1 word) — kebab-case flag names from the immediately following declarations, not actual descriptions.

## Proposed fix

Both bugs are one-line scorer changes.

**Bug 1 — three options, any single one suffices:**
1. Anchor `[^,]+` to non-newline: change to `[^,\n]+`. Kills cross-line capture entirely.
2. Change `{1,2}` to `{1,1}` since `XxxVar` always takes exactly four args (var ptr, name, default, usage). The `{1,2}` was presumably there to handle some pflag variant, but the regex's `(StringVar|IntVar|StringVarP|IntVarP)` allowlist all take four args.
3. Switch to `go/ast` parsing (correct but heavier).

Recommend option 1 (`[^,\n]+`) — minimal-risk one-character change.

**Bug 2 — word-boundary ID check:**

Replace:
```go
if strings.Contains(name, "id")
```
with:
```go
if name == "id" ||
    strings.HasPrefix(name, "id-") || strings.HasPrefix(name, "id_") ||
    strings.HasSuffix(name, "-id") || strings.HasSuffix(name, "_id") ||
    strings.Contains(name, "-id-") || strings.Contains(name, "_id_")
```

Keeps the intent (catch flags like `--id`, `--client-id`, `--video-id`, `--user_id`) without firing on `paid`, `idea`, `aided`, etc.

## Acceptance criteria

**Positive (multi-line capture fix):** A fixture with three consecutive `Flags().StringVar(&v, "flag-name-N", "", "Documented purpose for this flag")` declarations returns three 5-word descriptions, not artifact next-flag-names. Average word count = 5, not 3.

**Positive (ID word boundary fix):**
- `Flags().IntVar(&priceCents, "price-paid-cents", 0, "Price in cents")` does NOT contribute to `totalIDFlags`.
- `Flags().StringVar(&clientID, "client-id", "", "Client identifier")` DOES contribute to `totalIDFlags` (real ID flag).

**Negative (ID word boundary fix):** `Flags().IntVar(&itemID, "id", 0, "Item identifier")` still counts as an ID flag of wrong type (real ID flag, IntVar instead of StringVar, correctly flagged).

## Anti-pattern this fix prevents

Coffee-goat's session 3 had to rename `--price-paid-cents` → `--price-cents` to dodge the false positive — a real CLI surface change driven by a scorer bug. Renaming user-facing flags shouldn't be the per-CLI workaround for a scorer regex.

## Related issues

- #1255 — `verify-skill` has a separate regex-coverage bug in the same general territory (flag-declaration pattern matching). Different file, different tool, related-area only.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by vinnyp**
## Intent

`scoreTypeFidelity` in `internal/pipeline/scorecard.go` had three independent defects that systematically penalised SKILL-compliant CLIs. This PR fixes all three in one place, with dedicated regression tests for each.

Issue: Closes #1720, Closes #1559

## Approach

1. **`flagDeclRe` over-matched across newlines** (#1720 Bug A) — `[^,]+` matches newlines, so consecutive `cmd.Flags()....` calls bled into each other and the first call's description capture pulled in the next call's flag-name token. Replaced with `[^,\n]+` to bound each capture to a single Go statement.

2. **ID-flag substring false-positive** (#1720 Bug B) — `strings.Contains(name, "id")` matched any flag name containing `id` mid-word, including `price-paid-cents` (matched on the `id` inside `paid`). Extracted `isIDFlagName` with kebab-case word boundaries (equal to `id`, prefix `id-`, suffix `-id`, or contains `-id-`).

3. **MarkFlagRequired reward conflicted with the SKILL** (#1559) — the scorer awarded `+1` for sample files containing `MarkFlagRequired` 3+ times, but the SKILL's Phase-3 build checklist explicitly forbids it on hand-authored novel commands because Cobra evaluates it before `RunE`, breaking the verifier's `--dry-run` probe. Removed the reward block; required-flag validation belongs inside `RunE`.

Tests pin each fix individually so any future regression fails loudly. Solutions doc added under `docs/solutions/logic-errors/` per repo convention.

## Scope

Primary area: scorer (`internal/pipeline/scorecard.go`).

Why this belongs in this repo: the bug is in the machine's scorer, not in any one printed CLI. Every freshly generated CLI is affected.

## Risk

Low. All three fixes can only *stop* false negatives/positives — the regex tightening can't credit fewer things than before (it only stops over-matching across line boundaries), the word-boundary check is strictly more conservative on what it classifies as an ID flag, and dropping the MarkFlagRequired reward removes a path no SKILL-compliant CLI could legitimately use.

## Output Contract

`scripts/golden.sh verify` passes unchanged (20/20). No golden fixture used `MarkFlagRequired` or hit the cross-line capture failure mode, so no golden scorecard moved.

## Verification

- `go test ./internal/pipeline/ -run 'TestScoreTypeFidelity|TestIsIDFlagName' -count=1 -v` — pass.
- `go test ./...` — pass.
- `scripts/golden.sh verify` — pass (20/20).
- `go vet ./...` — pass.
- `go fmt ./...` — clean.

The reviewer agent flagged one Important issue on the first cut: a comment on the new regex test was misleading about whether the fixture was load-bearing, when it in fact pinned the fix (pre-fix code scores 3, fixed code scores 4). Verified empirically and fixed in commit `06a28d0`, along with cleaning up the pre-existing `"scores ... high"` subtest whose `MarkFlagRequired` calls became inert with the reward gone.

## AI / Automation Disclosure

- [ ] No AI or automation was used
- [x] Human-reviewed: AI or automation was used, and a human reviewed the work for intent, fit, and obvious issues before submission
- [ ] AI-reviewed only: an AI agent reviewed the work, but no human reviewed it before submission
- [ ] Fully automated: generated and submitted without human review for this specific change

### CODE REVIEW

File: docs/solutions/logic-errors/scoretypefidelity-flag-decl-regex-and-mark-required-2026-05-22.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+- Remove the `MarkFlagRequired` reward (the `requiredRe`, the `requiredCount` accumulator, and the `>=3` block). Required-flag validation belongs inside `RunE` per the SKILL.
+- Cover each fix with dedicated tests in `scorecard_tier2_test.go`: `TestIsIDFlagName` (table-driven word-boundary cases including the `price-paid-cents` regression), `TestScoreTypeFidelity_FlagDeclRegexBoundedToOneLine` (consecutive `Flags()` calls don't pollute each other's capture), and `TestScoreTypeFidelity_DoesNotRewardMarkFlagRequired` (a fixture with three `MarkFlagRequired` calls scores the same as one with none).
+
+## Why It Works
+
+The regex constraint is the smallest change that bounds each capture to a single Go statement; `[^,\n]+` is already the convention used elsewhere in the scorer for flag-call parsing. The kebab-case predicate matches the actual naming convention generated CLIs use, so it has no false positives on real flag names. Dropping the `MarkFlagRequired` reward removes the direct scorer-versus-SKILL conflict; SKILL-compliant CLIs can still reach the 5/5 cap via the ID-string check, the description-length check, and the dummy-guard-absence check.
```
[the remark this answers is not in the document]
Comment (vinnyp) [reply docs/solutions/logic-errors/scoretypefidelity-flag-decl-regex-and-mark-required-2026-05-22.md]: Good catch — fixed in `bea8f34d`. Updated the solutions doc to describe the achievable max accurately (`+2 + +1 + +1 = 4`), updated the `TypeFidelity` field comment to record both the effective max (`0-4`) and the structural slot the tier rollup still allocates (`5`), and lowered the in-function clamp from `> 5` to `> 4` so the dead branch goes away. The new regression tests already pin equality to `4`.

Reconciling the tier-rollup `tier2Max` constant (currently `60`) with the dim's 4-point achievable max is a separate concern — it changes the percentage for every printed CLI and needs a coordinated golden refresh, which is out of scope here.
